From 5830603c3ba7ebf0ef44a27f4fdab49a2397079f Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 28 Aug 2014 14:24:27 -0400
Subject: [cpufreq] intel_pstate: Disable interrupts during MSRs reading

Message-id: <1409235879-1283-174-git-send-email-prarit@redhat.com>
Patchwork-id: 89482
O-Subject: [RHEL7.1 PATCH BZ 1134639 173/185] cpufreq: intel_pstate: Disable interrupts during MSRs reading
Bugzilla: 1134369
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1134639

commit 4ab60c3f32c721e46217e762bcd3e55a8f659c04
Author: Stratos Karafotis <stratosk@semaphore.gr>
Date:   Fri Jul 18 08:37:24 2014 -0700

    cpufreq: intel_pstate: Disable interrupts during MSRs reading

    According to Intel 64 and IA-32 Architectures SDM, Volume 3,
    Chapter 14.2, "Software needs to exercise care to avoid delays
    between the two RDMSRs (for example interrupts)".

    So, disable interrupts during reading MSRs IA32_APERF and IA32_MPERF.
    This should increase the accuracy of the calculations.

    Signed-off-by: Stratos Karafotis <stratosk@semaphore.gr>
    Signed-off-by: Dirk Brandewie <dirk.j.brandewie@intel.com>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/cpufreq/intel_pstate.c b/drivers/cpufreq/intel_pstate.c
index b59b26b..a874890 100644
--- a/drivers/cpufreq/intel_pstate.c
+++ b/drivers/cpufreq/intel_pstate.c
@@ -569,9 +569,12 @@ static inline void intel_pstate_calc_busy(struct cpudata *cpu)
 static inline void intel_pstate_sample(struct cpudata *cpu)
 {
  u64 aperf, mperf;
+ unsigned long flags;
 
+ local_irq_save(flags);
  rdmsrl(MSR_IA32_APERF, aperf);
  rdmsrl(MSR_IA32_MPERF, mperf);
+ local_irq_restore(flags);
 
  aperf = aperf >> FRAC_BITS;
  mperf = mperf >> FRAC_BITS;
-- 
1.7.1